server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamMetadata.java

```diff
@@ -34,69 +33,70 @@ public class DataStreamMetadata implements Metadata.Custom {
             while (p.nextToken() != XContentParser.Token.END_OBJECT) {
                 String name = p.currentName();
                 dataStreams.put(name, DataStream.fromXContent(p));
             }
-            return dataStreams;
+            return dataStreams.build();
         }, DATA_STREAM);
         PARSER.declareObject(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> {
-            Map<String, DataStreamAlias> dataStreams = new HashMap<>();
+            ImmutableOpenMap.Builder<String, DataStreamAlias> dataStreams = ImmutableOpenMap.builder();
             while (p.nextToken() != XContentParser.Token.END_OBJECT) {
                 DataStreamAlias alias = DataStreamAlias.fromXContent(p);
                 dataStreams.put(alias.getName(), alias);
             }
-            return dataStreams;
+            return dataStreams.build();
         }, DATA_STREAM_ALIASES);
     }
 
-    private final Map<String, DataStream> dataStreams;
-    private final Map<String, DataStreamAlias> dataStreamAliases;
+    private final ImmutableOpenMap<String, DataStream> dataStreams;
+    private final ImmutableOpenMap<String, DataStreamAlias> dataStreamAliases;
 
-    public DataStreamMetadata(Map<String, DataStream> dataStreams, Map<String, DataStreamAlias> dataStreamAliases) {
-        this.dataStreams = Map.copyOf(dataStreams);
-        this.dataStreamAliases = Map.copyOf(dataStreamAliases);
+    public DataStreamMetadata(
+        ImmutableOpenMap<String, DataStream> dataStreams,
+        ImmutableOpenMap<String, DataStreamAlias> dataStreamAliases
+    ) {
+        this.dataStreams = dataStreams;
+        this.dataStreamAliases = dataStreamAliases;
     }
 
     public DataStreamMetadata(StreamInput in) throws IOException {
         this(
-            in.readImmutableMap(StreamInput::readString, DataStream::new),
-            in.readImmutableMap(StreamInput::readString, DataStreamAlias::new)
+            in.readImmutableOpenMap(StreamInput::readString, DataStream::new),
+            in.readImmutableOpenMap(StreamInput::readString, DataStreamAlias::new)
         );
     }
 
     public DataStreamMetadata withAddedDatastream(DataStream datastream) {
         final String name = datastream.getName();
         final DataStream existing = dataStreams.get(name);
-        if (existing == null) {
-            return new DataStreamMetadata(Maps.copyMapWithAddedEntry(dataStreams, name, datastream), dataStreamAliases);
-        } else if (existing.equals(datastream)) {
+        if (datastream.equals(existing)) {
             return this;
         }
-        return new DataStreamMetadata(Maps.copyMapWithAddedOrReplacedEntry(dataStreams, name, datastream), dataStreamAliases);
+        return new DataStreamMetadata(ImmutableOpenMap.builder(dataStreams).fPut(name, datastream).build(), dataStreamAliases);
     }
 
-    public Map<String, DataStream> dataStreams() {
+    public ImmutableOpenMap<String, DataStream> dataStreams() {
```

Comment:
Can the return type remain `Map`? I think we should at leat try to make `ImmutableOpenMap` an implementation detail of this class.

Reply:
I did it like this for now because we need to update the map in some of the methods in `Metadata`. I agree though, we should do something via copy-constructors here and just make the whole handling of this map internal to this class like we already did for adding a data stream. Maybe ok to push that into a follow-up to keep this change isolated?

Reply:
> Maybe ok to push that into a follow-up to keep this change isolated?

If we do this in a follow up change then I'm good with this change.

Reply:
I'll handle it, promise :)